examples: add example for adrv9002

nunojsa opened the pull request:
This is analogous to adrv9009-iiostream. Just serves as an example on how
to stream data on adrv9002 using libiio.

Signed-off-by: Nuno Sá <nuno.sa@analog.com>

pcercuei commented on examples/adrv9002-iiostream.c:
```diff
@@ -0,0 +1,322 @@
+		for (p_dat = iio_buffer_first(rx, rx_i_chan); p_dat < p_end; p_dat += p_inc) {
+			/* Example: swap I and Q */
+			int16_t i = get_unaligned(p_dat);
+			int16_t q = get_unaligned(&p_dat[2]);
```
Why the unaligned access?

nunojsa replied:
I took this from [here](https://github.com/analogdevicesinc/libiio/blob/master/examples/adrv9009-iiostream.c#L269). On that code we might clearly have some misaligned access since we have no guarantees that p_dat sits on a address aligned with a short.  The compiler will also complain about that code (using `-Wcast-align`)...

If you do not want to have these helpers, plain `memcpy()` should also make the compiler happy. Or if you can guarantee (as you know the libiio internals better than me :)) that those addresses will always have some alignment guarantees, I can just use the cast as in the adrv9009 example (though I think that is a bad practice).

pcercuei replied:
The buffers and the samples are guaranteed to be aligned.

If you make `p_dat` a `int16_t *`, then the compiler won't complain and you can just read i/q without casting the pointer.

nunojsa replied:
Ok, that works as long as I also do the update like `p_dat += p_inc / 2`... I guess p_inc is the complete size of the enabled scan elements?

pcercuei replied:
Yes, `p_inc` is the sample size in bytes.
You can update with: `p_dat += p_inc / sizeof(*p_dat)` just so that it's obvious what it does.

nunojsa commented:
V2:

* Drop unnecessary usage of `{get/put}_unaligned()` helpers.

tfcollins commented on examples/adrv9002-iiostream.c:
```diff
@@ -0,0 +1,304 @@
+			 * Example: fill with zeros
+			 * 14-bit sample needs to be MSB aligned so shift by 2
+			 * https://wiki.analog.com/resources/eval/user-guides/ad-fmcomms2-ebz/software/basic_iq_datafiles#binary_format
+			 */
```
Remove comment copied from ad9361-iiostream

nunojsa commented:
V3:

* Removed copied pasted comment.

pcercuei commented on examples/adrv9002-iiostream.c:
```diff
@@ -0,0 +1,299 @@
+	printf("%s, %d: "fmt, __func__, __LINE__, ##__VA_ARGS__)
+
+#define info(fmt, ...) \
+	printf("%s, %d: "fmt, __func__, __LINE__, ##__VA_ARGS__)
```
OSX doesn't like your `##__VA_ARGS__`...

nunojsa replied:
Arghhh, I guess I will also have to drop my neat cleanup handling?

https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=4022&view=logs&j=afc9f219-4f50-5293-48d5-d83ea574007e&t=470c08d8-8808-547a-c046-cb3882e7f5bd&l=185

pcercuei replied:
Shouldn't you use `atexit()`?

nunojsa replied:
Hmm the reason why I did not used it is because I cannot pass any argument to the cb... But for this example I can always make the objects global and that solves it! I will update it together with the variadic macro...

nunojsa replied:
Well, it actually makes it very cumbersome with `atexit()`. I will just drop this and add a normal cleanup function...

pcercuei replied:
Have a look at iio-debug.h in the dev branch, there are macros that basically do the same thing and are cross-platform.

nunojsa replied:
Just gave it a look but I do not think I can use it directly without some tweaks. The thing is that on the format I have, I want to prepend `__func__` and `__LINE__` to my format string. So, I cannot guarantee that `__VA_ARGS__` will expand to something as you can [here](https://github.com/analogdevicesinc/libiio/blob/dev/iio-debug.h#L85) with `buf`. First thing that comes to my mind is to just add a dummy empty string at the end of my format so that I can nicely extract my arguments. The other one is just using two printfs like:

```
#define error(...) \
        printf("%s, %d: ERROR: ", __func__, __LINE__); \
        printf(__VA_ARGS__);
```
Anyways, I just don't want to wast to much time on this... As I don't have HW with me today I will check this tomorrow. I want to do a sanity check before pushing this update...

pcercuei replied:
```
#define __error(line, ...) printf(__func__ ", " #line __VA_ARGS__);
#define _error(line, ...) __error(line, __VA_ARGS__);
#define error(fmt, ...) _error(__LINE__, fmt, __VA_ARGS__);
```

That doesn't work?

nunojsa replied:
Nope, I made a quick test and it does not compile. I'm not sure we can have `printf(__func__"plus something")`... What I have in mind for the dummy string is something like:

```
#define ARGS(fmt, ...)  __VA_ARGS__
#define FMT(fmt, ...)   fmt
#define error(...) \
       printf("%s, %d: ERROR: " FMT(__VA_ARGS__)"%s", __func__, __LINE__, ARGS(__VA_ARGS__, ""))
```

I could also just use `\n` instead of the empty string. I do not like much to enforce that but as this is an example app I don't really mind (if you prefer it...)

pcercuei replied:
@nunojsa 
```
#define __FIRST(a, ...)		a
#define ___OTHERS(a, ...)	a, __VA_ARGS__
#define __OTHERS(a, b, ...)	___OTHERS(a, __VA_ARGS__)
#define error(...) \
	printf("%s, %d: ERROR: " __FIRST(__VA_ARGS__, 0), __func__, __OTHERS(__LINE__,__VA_ARGS__))
```

nunojsa replied:
Does not works for all cases:

```
int n = 10;
error("Hello world: %d\n", n); // works fine
error("Hello world\n"); //compilation error
```
I think the issue with the later is that __VA_ARGS__ will evaluate to nothing in `___OTHERS` which is not nicely handled unless we use the gnu extension. That is why I think the "dummy" string after the format is needed (at least is one way to handle it; maybe not the best one).

Out of curiosity, is there any special reason for you to use `__FIRST(__VA_ARGS__, 0)` (also saw it in iio-debug.h)? If I'm not mistaken, `__FIRST(__VA_ARGS__)` should also work...

pcercuei replied:
C requires that the number of parameters in a macro's "..." is >= 1. iio-debug.h does `__FIRST(__VA_ARGS__, 0)` otherwise the compiler complains.

pcercuei replied:
Following your suggestion:
```
#include <stdio.h>

#define __FIRST(a, ...)		a
#define ___OTHERS(a, ...)	a, __VA_ARGS__
#define __OTHERS(a, b, ...)	___OTHERS(a, __VA_ARGS__)

#define error(...) \
	printf("%s, %d: ERROR: " __FIRST(__VA_ARGS__, 0) "%c", __func__, __OTHERS(__LINE__,__VA_ARGS__, '\0'))

int main(void)
{
	error("Test: %d\n", 5);
	error("Test2\n");
}
```

This seems to compile fine.

nunojsa replied:
> C requires that the number of parameters in a macro's "..." is >= 1. iio-debug.h does __FIRST(__VA_ARGS__, 0) otherwise the compiler complains.

Hmm, I see. I guess with gnu extensions this is not really the case. That is why gcc did not complained when I tried it...

pcercuei commented on examples/adrv9002-iiostream.c:
```diff
@@ -0,0 +1,299 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
```
You can strip that wall of text and use a SPDX license header, have a look at the C files in the top directory.

pcercuei reviewed (changes requested):
Just a couple of changes needed, right now it fails to build on OSX (see the autobuilders)

nunojsa commented:
V4:

* Dropped `on_exit()` and added a normal cleanup function;
* Refactored the log macros to be portable (no use of gnu ##VA_ARGS);
* `sigaction()` is part of POSIX and Windows is not fully compliant with it. Hence, use windows API for handling signals;
* When using `sigaction()`, make sure that the registered signals are unblocked...

nunojsa commented:
V5:

* Use SPDX license header...
